### ISSUE TITLE
Verify TypeMerge and TypeBranch instructions

### ISSUE THREAD

**Issue by manuranga** [ballerina-platform/nballerina#1022]
Following needs to be verified

Flow related:
* Non-backwards branches form a DAG
~~* Backward edges points to an entry node of the current node's [strongly connected component](https://en.wikipedia.org/wiki/Strongly_connected_component).
(Is this too strong? Will we ever have optimizations that violate this? This is a prerequisite for LOOP_REGION and reducible CFG) 
I plan to do this via [Tarjan's algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm).~~ This will be move to Region validation.
* Each Register is assigned before it is used.
* Check type flow :
`TypeBranch`'s `ifTrueRegister` can only be used in successors (inclusive) of `ifTrue`
`TypeBranch`'s `ifFalseRegister` can only be used in successors (inclusive)  of `ifFalse`
* Branches into a TypeMerge matches with its predecessors array

Local:
* TypeBranch: `operand  == rootUnderlying(ifTrueRegister) == rootUnderlying(ifFalseRegister)`
`operand.semType ∩ semType <: ifTrueRegister.semType` 
`operand.semType ∖ semType <: ifFalseRegister.semType` 
* `rootUnderlying(result) == rootUnderlying(operands[n])`
TypeMerge: `∪ operands[n].semType <: result.semType`

Impl choices:
* When to do the local checks (ie call `verifyBasicBlock`)
  - In a separate loop
  - During flow checking

* How to represent Initialized variables. This needs to pass down with the flow, bifurcating at `TypeBranch`.
  - Linked list similar to Bindings.
  - boolean array of size code.registers.length() 
    - Clone the array between sibling branches
    - Mark, unmark and reused for sibling branch

**Comment by jclark** [comment]
I think the easiest is a boolean array and clone between branches.

How are you planning to deal with something like:

```
if foo() {
   int x = 17;
   bar(x);
}
else {
   int y = 20;
   bar(y);
}
```

When you merge these branches, you have x initialized on one branch, and y initialized on the other branch.  They are  both out of scope, but you don't know that from the BIR (unless you use the debug scope info).

I *think* for out current subset we can treat them as both uninitialized.  But this obviously won't be enough for supporting local variable declarations without initializers. For this, I suggest you look at how the Java spec (and maybe C# spec) specify definite assignment.

**Comment by manuranga** [comment]
Plan is to allow Register access only when all paths to a usage has an assignment. But I assumed in the current subset all the merges that matter can be detected by having TypeMerge as the block's first insn. Plan was to create a merged initialized list at that point.

But your example shows I need to do it genericaly, not just at TypeMerge. Let me come up with a design.

**Comment by jclark** [comment]
> Plan is to allow Register access only when all paths to a usage has an assignment.

In our current subset, I don't think it is possible to have an variable that is for a block

- in scope in the block
- initialized on one path coming into the block
- not initialized on another path coming into the block

**Comment by manuranga** [comment]
> I don't think it is possible to have an variable…

Yes, it is not possible for variables, but it is possible for NarrowRegisters. So I think I need to generalize that logic to handle your example.

**Comment by jclark** [comment]
> Yes, it is not possible for variables, but it is possible for NarrowRegisters.

When there's a narrowed register, why won't it be initialized on both paths (by the TypeBranch)?

**Comment by manuranga** [comment]
Registers are typed with the occurrence type, so we have separate registers for true path and false path, each is only initialized on one side and should be invalidated after merge.

**Comment by jclark** [comment]
Still need to write some negative tests.

This waits on being able to read serialized BIR format.

**Comment by jclark** [comment]
Handle this in #1172 instead.
